[DEV-71] 티켓팅 전체 조회 EP 작성

dla0510 opened the pull request:
- 작업 내용
  - TicketingController, TicketingService 수정 및 관련 DTO 작성
  - TicketingServiceTest 추가 작성

dla0510 committed:
티켓팅 전체 조회 EP 작성

dla0510 committed:
Merge branch 'feat/DEV-72' of https://github.com/Tiketeer/Tiketeer-BE into feat/DEV-71

One-armed-boy commented on src/main/java/com/tiketeer/Tiketeer/domain/ticketing/dto/GetAllTicketingsDto.java:
```diff
@@ -0,0 +1,53 @@
+
+@ToString
+@Getter
+public class GetAllTicketingsDto {
+	private final UUID ticketingId;
+	private final String title;
+	private final String description;
+	private final String location;
+	private final String category;
+	private final Integer runningMinutes;
+	private final Integer stock;
+	private final Integer remainedStock;
+	private final Long price;
+	private final LocalDateTime eventTime;
+	private final LocalDateTime saleStart;
+	private final LocalDateTime saleEnd;
+	private final LocalDateTime createdAt;
```
티케팅 리스트에 요게 다 필요한거 맞나요? API Docs에도 전부 명시되어 있긴 하던데 진짜 필요한건지 의문...

dla0510 replied:
description, stock 정도는 빠져도 될 듯?
나머지는 FE 측에서 사용될 여지가 있다고 보입니다.
(현재 판매중인 티켓팅과 판매 예정인 티켓팅을 나눠서 보여주거나, remainedStock이 0이면 매진 등으로 표시하기 등)

dla0510 replied:
일단 description, stock는 빼뒀습니다
나머지는 낼 의논해보시죠

One-armed-boy commented on src/main/java/com/tiketeer/Tiketeer/domain/ticketing/service/TicketingService.java:
```diff
@@ -41,6 +43,33 @@ public TicketingService(TicketingRepository ticketingRepository, TicketService t
 		this.memberRepository = memberRepository;
 	}
 
+	@Transactional
+	public List<GetAllTicketingsDto> getAllTicketings() {
+		var ticketings = ticketingRepository.findAll()
```
조회만 할 때는 Transactional(readonly = true) 로 주시는게 성능 상 이점이 있습니다.

dla0510 committed:
@Transactional readonly true 설정

dla0510 committed:
전체 티켓 조회 반환 타입에서 description, stock 제외

dla0510 committed:
전체 티켓 조회 로직 수정

- ticketing_id로 모든 티켓을 조회하여 구매되지 않은 티켓만 필터링하여 통계내는 로직을 아래와 같이 수정
- ticketing_id와 purchase_id==null 조건을 줘서 티켓 조회 후 통계

One-armed-boy commented on src/main/java/com/tiketeer/Tiketeer/domain/ticket/repository/TicketRepository.java:
```diff
@@ -17,5 +17,7 @@ public interface TicketRepository extends JpaRepository<Ticket, UUID> {
 
 	List<Ticket> findAllByPurchase(Purchase purchase);
 
+	List<Ticket> findByTicketingIdAndPurchaseIsNullOrderById(UUID ticketingId);
+
```
혹시 왜 ID로 정렬시키는지 알 수 있을까요? 현재는 UUID V4가 들어가 있어서 정렬이 아무런 의미가 없습니다.

dla0510 replied:
아 limit 들어갈 때 순서가 항상 동일하게 출력되도록 하려고 order 붙이던 건데 여기는 빼도 될 것 같네요!
시간순 정렬의 의도가 있는 건 아닙니다

One-armed-boy replied:
어차피 PK라서 정렬해줘도 DB 단에서 추가적인 정렬 연산을 하지는 않을 것 같긴한데 왜 하셨나 궁금했습니다 ㅋㅋ

dla0510 replied:
흠.. 예전에 테이텀에서 order by 따로 안붙여주면 요청 보낼때마다 순서가 섞였던 기억이 있는데... 아닌가요?

dla0510 replied:
다른거랑 착각해서 잘못 기억한 걸수도ㅋㅋ

One-armed-boy replied:
쿼리 실행계획 까보시면 Sort 연산 없을꺼에요

One-armed-boy commented on src/main/java/com/tiketeer/Tiketeer/domain/ticketing/dto/GetAllTicketingsDto.java:
```diff
@@ -0,0 +1,47 @@
+import lombok.Getter;
+import lombok.ToString;
+
+@ToString
+@Getter
+public class GetAllTicketingsDto {
+	private final UUID ticketingId;
+	private final String title;
+	private final String location;
```
suffix가 ResponseDto가 되어야 하지 않나요?

dla0510 replied:
이 타입을 현재 TicketingService와 TicketingController에서 반환 타입으로 사용하고 있어서 suffix를 뺐습니다.

One-armed-boy replied:
서비스 Dto랑 컨트롤러 Dto 따로 분리하기로 했었지 않나여??

dla0510 replied:
그냥 똑같이 List로 반환하기로 한거 아니었나요??

dla0510 replied:
<img width="409" alt="image" src="https://github.com/Tiketeer/Tiketeer-BE/assets/43336212/fda66c7c-984b-4e22-a6f1-28c611671e0f">

여기서 2번으로 가고 Get이랑 List에 대해서만 작명 다르게 하자고 한 걸로 이해했습니다.

dla0510 replied:
어차피 낼 다시 논의하기로 했으니까 정리되면 수정하겠습니다!

One-armed-boy replied:
그럼 머지부터 하시나여? 아니면 수정되고 한번에 하실건가여

dla0510 replied:
펜딩되는 PR이 있어서 머지하고 나중에 한꺼번에 하겠슴다

One-armed-boy commented on src/main/java/com/tiketeer/Tiketeer/domain/ticketing/service/TicketingService.java:
```diff
@@ -32,13 +35,39 @@ public class TicketingService {
+		this.ticketRepository = ticketRepository;
+	}
+
+	@Transactional(readOnly = true)
+	public List<GetAllTicketingsDto> getAllTicketings() {
+		var ticketings = ticketingRepository.findAll()
+			.stream()
+			.map((ticketing) -> {
+				var remainedTickets = ticketRepository.findByTicketingIdAndPurchaseIsNullOrderById(
+					ticketing.getId());
+				return GetAllTicketingsDto.builder().ticketingId(ticketing.getId())
+					.price(ticketing.getPrice())
+					.category(ticketing.getCategory())
+					.location(ticketing.getLocation())
+					.title(ticketing.getTitle())
+					.runningMinutes(ticketing.getRunningMinutes())
+					.eventTime(ticketing.getEventTime())
+					.saleStart(ticketing.getSaleStart())
+					.saleEnd(ticketing.getSaleEnd())
+					.createdAt(ticketing.getCreatedAt())
+					.remainedStock(remainedTickets.size())
+					.build();
+			})
+			.toList();
+		return ticketings;
```
NL 조인을 로직 자체에서 구현한 느낌이라 성능이 너무 떨어질 것 같습니다...!

dla0510 replied:
흠... 성능은 차차 수정해보겠습니다

One-armed-boy replied:
그럼 TODO 주석만 추가해주시면 좋을듯

dla0510 replied:
요 부분도 추후에 querydsl 적용하면서 수정해보겠습니다

dla0510 replied:
주석은 추가해뒀습니다!

dla0510 committed:
리뷰 반영
